gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,505 @@
+          closeContentChannel();
+        } else {
+          seekDistance -= bytesRead;
+          contentChannelPosition += bytesRead;
+        }
+      } catch (IOException e) {
+        logger.atInfo().withCause(e).log(
+            "Got an IO exception on contentChannel.read(), a lazy-seek will be pending for '%s'",
+            resourceId);
+        closeContentChannel();
+      }
+    }
+    checkState(
+        contentReadChannel == null || contentChannelPosition == currentPosition,
+        "contentChannelPosition (%s) should be equal to currentPosition (%s)"
+            + " after successful in-place skip",
+        contentChannelPosition,
+        currentPosition);
+  }
+
+  private void resetContentChannel() {
+    checkState(contentReadChannel == null, "contentChannel should be null for '%s'", resourceId);
+    contentChannelPosition = -1;
+    contentChannelEnd = -1;
+  }
+
+  protected void closeContentChannel() {
+    if (contentReadChannel != null) {
+      logger.atFiner().log("Closing internal contentChannel for '%s'", resourceId);
+      try {
+        contentReadChannel.close();
+      } catch (Exception e) {
+        logger.atFine().withCause(e).log(
+            "Got an exception on contentChannel.close() for '%s'; ignoring it.", resourceId);
+      } finally {
+        contentReadChannel = null;
+        resetContentChannel();
+      }
+    }
+  }
+
+  private void performPendingSeeks(long bytesToRead) throws IOException {
+    if (currentPosition == contentChannelPosition && contentReadChannel != null) {
+      // No seek is required
+      return;
+    }
+
+    logger.atFiner().log(
+        "Performing lazySeek from %s to %s position for '%s'",
+        contentChannelPosition, currentPosition, resourceId);
+    // record the variation in channelPosition
+    // used to auto-detect random access
+    long oldPosition = contentChannelPosition;
+    long seekDistance = currentPosition - contentChannelPosition;
+    // Check if inPlace seek is a viable option
+    // If not close existing channel and open new one.
+
+    if (contentReadChannel != null
+        && seekDistance > 0
+        && seekDistance <= readOptions.getInplaceSeekLimit()
+        && currentPosition < contentChannelEnd) {
+      skipInPlace(seekDistance);
+    } else {
+      closeContentChannel();
+    }
+    if (contentReadChannel == null) {
+      if (isRandomAccessPattern(oldPosition)) {
+        setRandomAccess();
+      }
+      contentReadChannel = openReadChannel(bytesToRead);
+    }
+  }
+
+  // Update this function to create the reader and then set limits
+  private ReadableByteChannel getStorageReadChannel(long seek, long limit) throws IOException {
+    ReadChannel readChannel = storage.reader(blobId, generateReadOptions(blobId));
+    try {
+      readChannel.seek(seek);
+      // TODO: should we also add a check to not set limit if contentChannelEnd == objectSize?
+      if (limit != -1) {
```

Comment:
when can this happen?

Reply:
Not via current code but from a function completeness  point of view added this check where we don't specify the limit of channel. Think of it as `byte-range:0-` where we don't even set the end. 
If it really hurts, can remove this but in my opinion it's harmless.

Reply:
If this condition is not expected, it is better to add an assert, rather than handling this case.

Reply:
addressed.